File: server/go/session_store.go
package main

import (
	"sync"
	"time"

	glide "github.com/GlideIdentity/glide-be-sdk-go"
)

// SessionStore stores status_url from prepare responses for the polling proxy.
// This allows the status proxy to use the exact URL provided by the API.

type sessionEntry struct {
	StatusURL string
	ExpiresAt time.Time
}

var (
	sessionStore = make(map[string]sessionEntry)
	sessionMutex sync.RWMutex
)

func init() {
	// Cleanup expired sessions every minute
	go func() {
		ticker := time.NewTicker(time.Minute)
		for range ticker.C {
			cleanupExpiredSessions()
		}
	}()
}

func cleanupExpiredSessions() {
	sessionMutex.Lock()
	defer sessionMutex.Unlock()
	now := time.Now()
	for key, entry := range sessionStore {
		if entry.ExpiresAt.Before(now) {
			delete(sessionStore, key)
		}
	}
}

// StoreStatusURL stores a status URL for a session (5 minute TTL)
func StoreStatusURL(sessionKey, statusURL string) {
	sessionMutex.Lock()
	defer sessionMutex.Unlock()
	sessionStore[sessionKey] = sessionEntry{
		StatusURL: statusURL,
		ExpiresAt: time.Now().Add(5 * time.Minute),
	}
}

// GetStoredStatusURL retrieves the stored status URL for a session
func GetStoredStatusURL(sessionKey string) (string, bool) {
	sessionMutex.RLock()
	defer sessionMutex.RUnlock()
	entry, exists := sessionStore[sessionKey]
	if !exists || entry.ExpiresAt.Before(time.Now()) {
		return "", false
	}
	return entry.StatusURL, true
}

// ExtractStatusURL extracts status_url from a prepare response based on strategy
func ExtractStatusURL(response *glide.PrepareResponse) string {
	// SDK provides a helper function to extract status URL
	return glide.GetStatusURL(response)
}

File: server/node/session-store.ts
/**
 * Session Store for Status URLs
 * 
 * Stores status_url from prepare responses for the polling proxy.
 * This allows the status proxy to use the exact URL provided by the API.
 */

import type { PrepareResponse } from '@glideidentity/glide-be-sdk-node';
import { getStatusUrl as sdkGetStatusUrl } from '@glideidentity/glide-be-sdk-node';

interface SessionEntry {
  statusUrl: string;
  expiresAt: number;
}

const sessionStore = new Map<string, SessionEntry>();

// Cleanup expired sessions every minute
setInterval(() => {
  const now = Date.now();
  for (const [key, entry] of sessionStore.entries()) {
    if (entry.expiresAt < now) {
      sessionStore.delete(key);
    }
  }
}, 60 * 1000);

/** Store a status URL for a session (5 minute TTL) */
export function storeStatusUrl(sessionKey: string, statusUrl: string): void {
  sessionStore.set(sessionKey, {
    statusUrl,
    expiresAt: Date.now() + 5 * 60 * 1000,
  });
}

/** Get the stored status URL for a session */
export function getStatusUrl(sessionKey: string): string | undefined {
  const entry = sessionStore.get(sessionKey);
  if (entry && entry.expiresAt > Date.now()) {
    return entry.statusUrl;
  }
  if (entry) {
    sessionStore.delete(sessionKey);
  }
  return undefined;
}

/** Extract status_url from a prepare response based on strategy */
export function extractStatusUrl(response: PrepareResponse): string | undefined {
  // SDK provides a helper function to extract status URL
  return sdkGetStatusUrl(response);
}
